src/dune_engine/process.ml

```diff
@@ -572,6 +577,104 @@ end = struct
       fail ~loc ~annots paragraphs
 end
 
+type t =
+  { started_at : float
+  ; pid : Pid.t
+  ; response_file : Path.t option
+  ; stdout : Path.t option
+  ; stderr : Path.t option
+  ; stdout_on_success : Action_output_on_success.t
+  ; stderr_on_success : Action_output_on_success.t
+  }
+
+module Result = struct
+  type nonrec process = t
+
+  module Out = struct
+    type state =
+      | No_capture
+      | File of Path.t
+      | Read of string
+      | Closed
+
+    type t =
+      { on_success : Action_output_on_success.t
+      ; mutable unexpected_output : bool
```

Comment:
I don't like mutable fields here. For now, it's fine because this type is not exposed anywhere in the .mli, but it seems wrong to provide an interface with it.

Reply:
Yeah, it's a bit ugly, but I encourage you to understand the old code though :)

We handle 3 separate concerns with mutation:

1. We make sure that we never read the file unless we need the output
2. Subsequent reads are free
3. If we need to swallow the output, we make sure the output we read is always empty.

It also allows us to introduce an optimization:

To check if the output is empty, we don't need to read the file. We can just get its size using stat. That saves an open/close on empty files which should be quite common.